File: src/curate_ai/ingestion/arxiv.py
"""arXiv API scraper for the ingestion module."""

from datetime import datetime, timedelta, timezone

import feedparser
import httpx

from curate_ai.ingestion.base import BaseScraper, IngestionResult, SourceConfig
from curate_ai.logging import get_logger

logger = get_logger(__name__)


class ArxivFetcher(BaseScraper):
    """Fetcher for arXiv research papers."""
    
    API_URL = "https://export.arxiv.org/api/query"
    
    def __init__(self, config: SourceConfig):
        super().__init__(config)
        self.arxiv_config = config.arxiv
    
    async def fetch(self, days_back: int | None = None) -> list[IngestionResult]:
        """Fetch papers from arXiv API."""
        if not self.arxiv_config.get("enabled", True):
            logger.info("arXiv disabled")
            return []
        
        categories = self.arxiv_config.get("categories", ["cs.AI", "cs.LG", "cs.CL"])
        max_results = self.arxiv_config.get("max_results", 50)
        
        if days_back is None:
            days_back = self.arxiv_config.get("days_lookback", 3)
        
        # Build query
        cat_query = " OR ".join(f"cat:{cat}" for cat in categories)
        
        params = {
            "search_query": cat_query,
            "sortBy": "submittedDate",
            "sortOrder": "descending",
            "start": 0,
            "max_results": max_results,
        }
        
        try:
            async with httpx.AsyncClient(
                timeout=self.get_timeout(),
                follow_redirects=True,
            ) as client:
                response = await client.get(self.API_URL, params=params)
                response.raise_for_status()
        except Exception as e:
            logger.error("arXiv API failed", error=str(e))
            return []
        
        # Parse Atom feed
        feed = feedparser.parse(response.text)
        cutoff_date = datetime.now(timezone.utc) - timedelta(days=days_back)
        
        results = []
        for entry in feed.entries:
            # Parse publication date
            try:
                published = datetime.fromisoformat(
                    entry.published.replace("Z", "+00:00")
                )
                if published < cutoff_date:
                    continue
            except (AttributeError, ValueError):
                continue
            
            # Extract primary category
            primary_cat = entry.get("arxiv_primary_category", {}).get("term", "")
            
            # Extract all categories
            categories = [tag.term for tag in entry.get("tags", [])]
            
            # Extract authors
            authors = [author.name for author in entry.get("authors", [])]
            
            # Clean title and summary
            title = entry.title.replace("\n", " ").strip()
            summary = entry.summary.replace("\n", " ").strip()
            
            # Get PDF link
            pdf_link = entry.link
            for link in entry.get("links", []):
                if link.get("type") == "application/pdf":
                    pdf_link = link.get("href", entry.link)
                    break
            
            results.append(IngestionResult(
                title=title,
                url=entry.link,
                source="arXiv",
                source_type="arxiv",
                category="research",
                summary=summary[:1500],
                published_at=published,
                authors=authors,
                tags=categories,
                metadata={
                    "primary_category": primary_cat,
                    "pdf_url": pdf_link,
                    "arxiv_id": entry.id.split("/abs/")[-1] if "/abs/" in entry.id else entry.id,
                },
            ))
        
        logger.info("arXiv fetch completed", count=len(results), categories=len(categories))
        return results

File: packages/backend/src/ingestion/arxiv.ts
import { XMLParser } from "fast-xml-parser";
import type { ArxivConfig } from "./sources.js";
import type { RawItem } from "./types.js";

interface AtomEntry {
  title?: string;
  summary?: string;
  published?: string;
  id?: string;
  author?: unknown;
  link?: unknown;
  category?: unknown;
  "arxiv:primary_category"?: unknown;
  "arxiv:comment"?: unknown;
}

function toArray<T>(v: T | T[] | undefined): T[] {
  if (v === undefined) return [];
  return Array.isArray(v) ? v : [v];
}

function str(v: unknown): string {
  if (typeof v === "string") return v;
  if (v && typeof v === "object") {
    const o = v as { "#text"?: string };
    if (typeof o["#text"] === "string") return o["#text"];
  }
  return "";
}

function attr(v: unknown, name: string): string | undefined {
  if (v && typeof v === "object") {
    const o = v as Record<string, unknown>;
    const val = o[`@_${name}`];
    if (typeof val === "string") return val;
  }
  return undefined;
}

export interface ArxivFetchResult {
  items: RawItem[];
  error?: string;
}

/** Fetch recent papers from the arXiv Atom API for the configured categories. */
export async function fetchArxiv(
  config: ArxivConfig,
  maxResults = 30,
  daysBack = 7,
  fetcher: typeof fetch = fetch,
): Promise<ArxivFetchResult> {
  if (!config.enabled) return { items: [] };
  try {
    const catQuery = config.categories.map((c) => `cat:${c}`).join(" OR ");
    const params = new URLSearchParams({
      search_query: catQuery,
      sortBy: "submittedDate",
      sortOrder: "descending",
      start: "0",
      max_results: String(Math.min(maxResults, 200)),
    });
    const res = await fetcher(`https://export.arxiv.org/api/query?${params}`, {
      headers: { "User-Agent": "CurateAI/0.2 (research curation)" },
    });
    if (!res.ok) return { items: [], error: `arXiv HTTP ${res.status}` };
    const xml = await res.text();

    const parser = new XMLParser({ ignoreAttributes: false });
    const doc = parser.parse(xml) as { feed?: { entry?: AtomEntry | AtomEntry[] } };
    const entries = toArray<AtomEntry>(doc.feed?.entry);

    const cutoff = Date.now() - daysBack * 86_400_000;
    const items: RawItem[] = [];

    for (const entry of entries) {
      const published = entry.published ?? "";
      const publishedMs = Date.parse(published);
      if (Number.isNaN(publishedMs)) continue;
      if (publishedMs < cutoff) continue;

      const title = str(entry.title).replace(/\s+/g, " ").trim();
      const summary = str(entry.summary).replace(/\s+/g, " ").trim();
      const id = str(entry.id);
      const absUrl = id.replace(/^http:/, "https:");
      const authors = toArray(entry.author)
        .map((a) => str((a as { name?: string })?.name))
        .filter((a): a is string => Boolean(a));
      const links = toArray(entry.link);
      let pdfUrl = absUrl;
      for (const link of links) {
        if (attr(link, "type") === "application/pdf") {
          pdfUrl = attr(link, "href") ?? absUrl;
          break;
        }
      }
      const categories = toArray(entry.category)
        .map((c) => attr(c, "term"))
        .filter((c): c is string => Boolean(c));
      const primary =
        attr(entry["arxiv:primary_category"], "term") ?? categories[0] ?? "";
      const arxivId = absUrl.split("/abs/").pop() ?? "";

      items.push({
        title,
        url: absUrl,
        source: "arXiv",
        sourceType: "arxiv",
        summary,
        content: summary.slice(0, 3000),
        publishedAt: new Date(publishedMs).toISOString(),
        tags: categories,
        metadata: { primaryCategory: primary, pdfUrl, arxivId },
      });
    }

    return { items };
  } catch (err) {
    return { items: [], error: err instanceof Error ? err.message : String(err) };
  }
}
